File: LTRLib.Windows/Extensions/FormsExtensions.vb
'
' LTRLib
' 
' Copyright (c) Olof Lagerkvist, LTR Data
' http://ltr-data.se   https://github.com/LTRData
'

Imports System.Drawing
Imports System.Drawing.Imaging
Imports System.IO
Imports System.Runtime.CompilerServices
Imports System.Security
Imports System.Text
Imports System.Windows.Forms
#If NET30_OR_GREATER OrElse NETCOREAPP Then
Imports System.Windows.Media.Media3D
#End If
Imports LTRLib.IO

Namespace Extensions

    Public Module FormsExtensions

        <Extension>
        Public Function GetTopMostOwner(form As Form) As Form

            While form?.Owner IsNot Nothing
                form = form.Owner
            End While

            Return form

        End Function

        <Extension>
        Public Function ToBase64Url(img As Image) As String

            Using buffer As New MemoryStream()
                img.Save(buffer, ImageFormat.Png)
                Return DrawingSupport.PngImageToBase64Url(buffer.ToArray())
            End Using

        End Function

#If NET30_OR_GREATER OrElse NETCOREAPP Then

        ''' <summary>
        ''' Calculates the area covered by a size.
        ''' </summary>
        <Extension>
        Public Function Volume(Size As Size3D) As Double
            Return Size.X * Size.Y * Size.Z
        End Function

        <Extension>
        Public Function Radius(Point As Point3D) As Double
            Return Math.Sqrt(Math.Pow(Math.Sqrt(Point.X * Point.X + Point.Y * Point.Y), 2) + Point.Z * Point.Z)
        End Function

        <Extension>
        Public Function DistanceTo(Point1 As Point3D, Point2 As Point3D) As Double
            Return (Point1 - New Vector3D(Point2.X, Point2.Y, Point2.Z)).Radius()
        End Function

        <Extension>
        Public Function CenterPoint(Rectangle As Rect3D) As Point3D
            Return New Point3D(Rectangle.X + Rectangle.SizeX / 2, Rectangle.Y + Rectangle.SizeY / 2, Rectangle.Z + Rectangle.SizeZ / 2)
        End Function

#End If

        ''' <summary>
        ''' Calculates the area covered by a size.
        ''' </summary>
        <Extension>
        Public Function Area(Size As Size) As Integer
            Return Size.Height * Size.Width
        End Function

        <Extension>
        Public Function Radius(Point As Point) As Double
            Return Math.Sqrt(Point.X * Point.X + Point.Y * Point.Y)
        End Function

        ''' <summary>
        ''' Calculate a reverse color.
        ''' </summary>
        ''' <param name="OrigColor">Original color</param>
        ''' <returns>Reverse color.</returns>
        <Extension>
        Public Function InverseColor(OrigColor As Color) As Color
            Return Color.FromArgb(Not OrigColor.ToArgb())
        End Function

        ''' <summary>
        ''' Calculates the area covered by a size.
        ''' </summary>
        <Extension>
        Public Function Area(Size As SizeF) As Single
            Return Size.Height * Size.Width
        End Function
        <Extension>
        Public Function Radius(Point As PointF) As Double
            Return Math.Sqrt(Point.X * Point.X + Point.Y * Point.Y)
        End Function

        <Extension>
        Public Function DistanceTo(Point1 As Point, Point2 As Point) As Double
            Return (Point1 - New Size(Point2)).Radius()
        End Function

        <Extension>
        Public Function DistanceTo(Point1 As PointF, Point2 As PointF) As Double
            Return (Point1 - New SizeF(Point2)).Radius()
        End Function

        <Extension>
        Public Function Acos(Point As PointF) As Double
            Return Math.Acos(Point.Y / Point.X)
        End Function

        <Extension>
        Public Function Asin(Point As PointF) As Double
            Return Math.Asin(Point.Y / Point.X)
        End Function

        <Extension>
        Public Function Atan(Point As PointF) As Double
            Return Math.Atan2(Point.Y, Point.X)
        End Function

        <Extension>
        Public Function CenterPoint(Rectangle As Rectangle) As Point
            Return New Point(Rectangle.Left + Rectangle.Width \ 2, Rectangle.Top + Rectangle.Height \ 2)
        End Function

        <Extension>
        Public Function CenterPoint(Rectangle As RectangleF) As PointF
            Return New PointF(Rectangle.Left + Rectangle.Width / 2, Rectangle.Top + Rectangle.Height / 2)
        End Function

#If NETFRAMEWORK AndAlso Not NET45_OR_GREATER Then
        ''' <summary>
        ''' Adjusts a string to a display with specified width by wrapping complete words.
        ''' </summary>
        ''' <param name="Msg"></param>
        ''' <param name="LineWidth">Display width. If omitted, defaults to console window width.</param>
        ''' <param name="WordDelimiter">Word separator character.</param>
        ''' <param name="FillChar">Fill character.</param>
        <Extension, SecuritySafeCritical>
        Public Function LineFormat(Msg As String, Optional LineWidth As Integer? = Nothing, Optional WordDelimiter As Char = " "c, Optional FillChar As Char = " "c) As String

            Dim Width As Integer

            If LineWidth.HasValue Then
                Width = LineWidth.Value
            Else

                If Win32API.GetFileType(Win32API.GetStdHandle(NativeConstants.StdHandle.Output)) <> NativeConstants.Win32FileType.Char Then
                    Width = 79
                Else
                    Width = Console.WindowWidth - 1
                End If

            End If

            Dim origLines = Msg.Nz().Replace(vbCr, "").Split({vbLf(0)})

            Dim resultLines As New List(Of String)(origLines.Length)

            For Each origLine In origLines

                Dim Result As New StringBuilder

                Dim Line As New StringBuilder(Width)

                For Each Word In origLine.Split(WordDelimiter)
                    If Word.Length >= Width Then
                        Result.AppendLine(Word)
                        Continue For
                    End If
                    If Word.Length + Line.Length >= Width Then
                        Result.AppendLine(Line.ToString())
                        Line.Length = 0
                    End If
                    If Line.Length > 0 Then
                        Line.Append(WordDelimiter)
                    End If
                    Line.Append(Word)
                Next

                If Line.Length > 0 Then
                    Result.Append(Line)
                End If

                resultLines.Add(Result.ToString())

            Next

#If NET40_OR_GREATER Then
            Return String.Join(Environment.NewLine, resultLines)
#Else
            Return String.Join(Environment.NewLine, resultLines.ToArray())
#End If

        End Function
#End If

    End Module

End Namespace

File: LTRLib.Windows/Extensions/FormsExtensions.cs
// 
// LTRLib
// 
// Copyright (c) Olof Lagerkvist, LTR Data
// http://ltr-data.se   https://github.com/LTRData
// 

using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using System.Security;
using System.Text;
using System.Windows.Forms;

#if NET30_OR_GREATER || NETCOREAPP
using System.Windows.Media.Media3D;
#endif

using LTRLib.IO;

namespace LTRLib.Extensions;

public static class FormsExtensions
{
    public static Form? GetTopMostOwner(this Form? form)
    {
        while (form?.Owner is not null)
        {
            form = form.Owner;
        }

        return form;
    }

    public static string? ToBase64Url(this Image img)
    {
        using var buffer = new MemoryStream();
        img.Save(buffer, ImageFormat.Png);
        return DrawingSupport.PngImageToBase64Url(buffer.ToArray());
    }

    #if NET30_OR_GREATER || NETCOREAPP

    /// <summary>
    /// Calculates the area covered by a size.
    /// </summary>
    public static double Volume(this Size3D Size)
    {
        return Size.X * Size.Y * Size.Z;
    }

    public static double Radius(this Point3D Point)
    {
        return Math.Sqrt(Math.Pow(Math.Sqrt(Point.X * Point.X + Point.Y * Point.Y), 2d) + Point.Z * Point.Z);
    }

    public static double DistanceTo(this Point3D Point1, Point3D Point2)
    {
        return (Point1 - new Vector3D(Point2.X, Point2.Y, Point2.Z)).Radius();
    }

    public static Point3D CenterPoint(this Rect3D Rectangle)
    {
        return new Point3D(Rectangle.X + Rectangle.SizeX / 2d, Rectangle.Y + Rectangle.SizeY / 2d, Rectangle.Z + Rectangle.SizeZ / 2d);
    }

    #endif

    /// <summary>
    /// Calculates the area covered by a size.
    /// </summary>
    public static int Area(this Size Size)
    {
        return Size.Height * Size.Width;
    }

    public static double Radius(this Point Point)
    {
        return Math.Sqrt(Point.X * Point.X + Point.Y * Point.Y);
    }

    /// <summary>
    /// Calculate a reverse color.
    /// </summary>
    /// <param name="OrigColor">Original color</param>
    /// <returns>Reverse color.</returns>
    public static Color InverseColor(this Color OrigColor)
    {
        return Color.FromArgb(~OrigColor.ToArgb());
    }

    /// <summary>
    /// Calculates the area covered by a size.
    /// </summary>
    public static float Area(this SizeF Size)
    {
        return Size.Height * Size.Width;
    }
    public static double Radius(this PointF Point)
    {
        return Math.Sqrt((double)(Point.X * Point.X + Point.Y * Point.Y));
    }

    public static double DistanceTo(this Point Point1, Point Point2)
    {
        return (Point1 - new Size(Point2)).Radius();
    }

    public static double DistanceTo(this PointF Point1, PointF Point2)
    {
        return (Point1 - new SizeF(Point2)).Radius();
    }

    public static double Acos(this PointF Point)
    {
        return Math.Acos((double)(Point.Y / Point.X));
    }

    public static double Asin(this PointF Point)
    {
        return Math.Asin((double)(Point.Y / Point.X));
    }

    public static double Atan(this PointF Point)
    {
        return Math.Atan2((double)Point.Y, (double)Point.X);
    }

    public static Point CenterPoint(this Rectangle Rectangle)
    {
        return new Point(Rectangle.Left + Rectangle.Width / 2, Rectangle.Top + Rectangle.Height / 2);
    }

    public static PointF CenterPoint(this RectangleF Rectangle)
    {
        return new PointF(Rectangle.Left + Rectangle.Width / 2f, Rectangle.Top + Rectangle.Height / 2f);
    }

    #if NETFRAMEWORK && !NET45_OR_GREATER
    /// <summary>
    /// Adjusts a string to a display with specified width by wrapping complete words.
    /// </summary>
    /// <param name="Msg"></param>
    /// <param name="LineWidth">Display width. If omitted, defaults to console window width.</param>
    /// <param name="WordDelimiter">Word separator character.</param>
    /// <param name="FillChar">Fill character.</param>
    [SecuritySafeCritical]
    public static string LineFormat(this string Msg, int? LineWidth = default, char WordDelimiter = ' ', char FillChar = ' ')
    {

        int Width;

        if (LineWidth.HasValue)
        {
            Width = LineWidth.Value;
        }

        else if (IO.Win32API.GetFileType(IO.Win32API.GetStdHandle(NativeConstants.StdHandle.Output)) != NativeConstants.Win32FileType.Char)
        {
            Width = 79;
        }
        else
        {
            Width = Console.WindowWidth - 1;
        }

        var origLines = Msg.Nz().Replace("\r", "").Split('\n');

        var resultLines = new List<string>(origLines.Length);

        foreach (var origLine in origLines)
        {
            var Result = new StringBuilder();

            var Line = new StringBuilder(Width);

            foreach (var Word in origLine.Split(WordDelimiter))
            {
                if (Word.Length >= Width)
                {
                    Result.AppendLine(Word);
                    continue;
                }
                if (Word.Length + Line.Length >= Width)
                {
                    Result.AppendLine(Line.ToString());
                    Line.Length = 0;
                }
                if (Line.Length > 0)
                {
                    Line.Append(WordDelimiter);
                }
                Line.Append(Word);
            }

            if (Line.Length > 0)
            {
                Result.Append(Line);
            }

            resultLines.Add(Result.ToString());
        }

#if NET40_OR_GREATER

        return string.Join(Environment.NewLine, resultLines);

#else

        return string.Join(Environment.NewLine, resultLines.ToArray());

#endif
    }

#endif
}

